---
title: "Basic Use of ceRNAnetsim"
author: "Selcen Ari"
date: "`r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{basic_usage}
  %\VignetteEngine{knitr::rmarkdown}
  %\VignetteEncoding{UTF-8}
---

```{r, include = FALSE}
knitr::opts_chunk$set(
  collapse = TRUE,
  comment = "#>"
)
```

```{r setup, message= FALSE, warning=FALSE}
#install.packages("devtools")
#devtools::install_github("selcenari/ceRNAnetsim")
library(ceRNAnetsim)
```

### About the data

Below is the minimal data that can be used with ceRNAnetsim.

```{r}
minsamp %>% 
  select(1:4)
```


The table is actually constructed by merging three different tables:

* gene expression data
* miRNA expression data
* miRNA:gene pairs (might **optionally** contain data about the interaction)

So, the `basic_data` table is constructed by merging following tables:

```{r, message= FALSE, warning=FALSE, echo=FALSE}
minsamp %>% 
  select(competing, Competing_expression) %>% 
  distinct() -> gene_expression
```

```{r}
gene_expression
```

```{r, message= FALSE, warning=FALSE, echo=FALSE}
minsamp %>% 
  select(miRNA, miRNA_expression) %>% 
  distinct() -> mirna_expression
```

```{r}
mirna_expression
```

Third table should contain miRNA:gene interactions per row. The `ceRNAnetsim` will assume first column contains competing RNA names and second column to be miRNA names. If the order is different the user should indicate column names accordingly. 

```{r, message= FALSE, warning=FALSE, echo=FALSE}
minsamp %>% 
  select(competing, miRNA) -> interaction_simple
```


```{r}
interaction_simple
```

The three tables can be joined in R (as shown below) or elsewhere to have interaction and expression data altogether in expected format.

```{r}
interaction_simple %>%
  inner_join(gene_expression, by = "competing") %>%
  inner_join(mirna_expression, "miRNA") -> basic_data

basic_data
```



## What does ceRNAnetsim do?

ceRNAnetsim processes your dataset as graph object and simulates competing behaviours of targets when steady-state is perturbed via expression level changes in one or more genes. Let's go over three steps:

## 1. Handle basic dataset

In first step, the expression and interaction table is converted into graph/network. `tidygraph` is used importing the data thus both node and edge data are accessible as tables if needed. `priming_graph` generates many columns in edge/node table which are mostly for internal use.

```{r}
#Convertion of dataset to graph.

priming_graph(basic_data, competing_count = Competing_expression, miRNA_count =miRNA_expression)
```

## 2. Trigger a change

`update_how` function can be used to simulate a change in the network. (If multiple chnages are aimed to be used as trigger, `update_variables()` function should be used).

In the example below, expression level of "Gene2" is increased to two-fold.


```{r}

priming_graph(basic_data, competing_count = Competing_expression, 
              miRNA_count =miRNA_expression) %>%
  update_how(node_name = "Gene2", how=2)

```

You can see the current count of Gene2 node is 20000 and its change is denoted as "Up" in `changes_variable` column in node table data.

## 3. Simulate the changes in graph

Finally, with the help of `simulate` function, the effect of expression change (*i.e.* the trigger) on overall network. The example code advances only for 5 cycles.

```{r}

priming_graph(basic_data, competing_count = Competing_expression, 
              miRNA_count =miRNA_expression) %>%
  update_how(node_name = "Gene2", how=2) %>%
  simulate(cycle = 5)

```

`count_current` and `changes_variable` columns indicate results after 5 cycle of calculations for each node.

## A special case: knockdown

ceRNAnetsim also provides the simulation of gene knockdown in the network. In normal conditions, when a gene is up or down regulated, it is considered that amounts of gene transcripts change depended on interactions. But, the transcripts of the gene are not observed in the system when it is knocked down. To achieve this case, you just need to define `how` argument to 0 (zero) in `update_how` function. Note that, If `how` argument is specified as `0`, *simulate()* and *update_how()* functions process the variables to knockdown of specified gene with default `knockdown = TRUE`. For this reason, the workflow must be done with `knockdown= FALSE`, when it is desirable that the presence of transcripts of specified gene in the system.

```{r}
data("midsamp")

midsamp %>%
   priming_graph(competing_count = Gene_expression, 
                 miRNA_count = miRNA_expression, 
                 aff_factor = c(seeds, Energy), 
                 deg_factor = targeting_region) %>%
  update_how("Gene4", how=0) %>%
  simulate(cycle = 10)
```

Briefly, ceRNAnetsim utilizes the change(s) as trigger and calculates regulation of targets according to miRNA:target and target:total target ratios. 

See the other vignettes for more information.

